XXX-staged-require-with-lifecycle-hooks.md

```diff
@@ -0,0 +1,192 @@
+
+Each stage would interact with a `ModuleRequest` object which expresses the
+lifetime of a module request through the `resolve`, `load` and `compile` stages.
+Initially, it would only include the `parent` and `inputPath` fields. Other
+fields would be populated in the different lifecycle stages.
+
+### `ModuleRequest`
+
+##### `moduleRequest.parent`
+
+The parent module which initiated the `require()` request.
+
+##### `moduleRequest.inputPath`
+
+The input string to the `require()` call or `import` statement which initiated
+this module request.
+
+##### `moduleRequest.resolvedPath`
+
+The fully resolved file path, as determined by the `require.resolvers` handler.
+This would be added to the module request object during the `resolver` stage.
+
+##### `moduleRequest.contents`
+
+The content string (or buffer?) of the module. This field is populated during
+the `loader` stage.
+
+##### `moduleRequest.module`
+
+The final module object, built by the `compiler` stage.
+
+### `require.resolvers`
+
+It is the responsibility of `require.resolvers` extension handlers to receive
+the in-process `Module` object and use the metadata on it to figure out how to
+resolve a load. For `require()` relative location loading, it'd simply look at
+the `parent` and resolve the input string of the `require()` call relative to
+the `parent` module.
+
+This would roughly correspond to the current `Module._resolveLookupPaths`
+implementation, but it should instead receive the module request object which
+it will modify in-place.
+
+```js
+const mockedPaths = {
+  'foo.js': path.join(__dirname, 'mocked-foo.js')
+}
+
+const originalResolver = require.resolvers['require']
+require.resolvers['require'] = function customResolver(request) {
+  if (Object.keys(mockedPaths).includes(request.inputPath)) {
+    return mockedPaths[request.inputPath]
+  }
+  originalResolver(request)
+}
+```
+
+By defining named behaviour specific to `require()`, we open the door for
+supporting the different behaviour of `import` in the future.
+
+### `require.loaders`
+
+The `require.loaders` extension handlers receive the `Module` object after the
+`resolver` stage and use the resolved path to attempt to load the contents of
+the module file.
+
+```js
+const originalLoader = require.loaders['.js']
+require.loaders['.js'] = function customLoader(request) {
+  // This populates the `contents` field
+  originalLoader(request)
+
+  request.contents = request.contents.replace(
+    /module\.exports\s+\=/,
+    // Don't forget that `exports` object now!
+    'exports = module.exports ='
+  )
+}
+```
+
+### `require.compilers`
+
+The `require.compilers` extension handlers receive the `Module` object after the
+`loader` stage and pass the `contents` into a compiler function to build the
+final module object.
+
+```js
+require.compilers['.yaml'] = function yamlCompiler(request) {
+  request.module.exports = yaml.parse(request.contents)
+}
+```
```

Comment:
also seems like it should be a stack. If the code was .ts, it might need compilation to .js, but after it's compiled to js, an APM that instruments js would want to get it so it can be rewritten for instrumentation